samples/scanner/scanners/firewall_rules/matches.yaml

```diff
@@ -0,0 +1,114 @@
+# This file contains several example blacklist firewall rules.
+# 
+# Note: Required and Matches rules can only have "match_policies" fields which will
+# be validated.
+#
+# Matches rules are evaluated in the following steps:
+# 1. Get list of all pairs of resources and policies that apply to the resource.
+# 2. For each pair, get a list of applicable rules by:
+#   a. Checking if a rule applies to the specific resource.
+#   b. Checking each parent in the hierarchy for a rule that applies to that parent.
+#     See: https://cloud.google.com/resource-manager/docs/cloud-platform-resource-hierarchy
+#   c. Return the first applicable rules.
+# 3. Check if the policy matches all of the "match_policies" defined in the rule.
+#     If it does not, then this creates a RuleViolation.
+rules:
+
+    # This rule will trigger on policies that do not match the required policy.
+    #
+    # Example triggering firewall rules:
+    # [
+    #   {
+    #     "name": "firewall1",
+    #     "network": "https://www.googleapis.com/compute/v1/projects/yourproject/global/networks/default",
+    #     "priority": 1000,
+    #     "sourceRanges": [
+    #       "10.0.0.0/8",
+    #       ],
+    #     "allowed": [
+    #       {
+    #         "IPProtocol": "tcp",
+    #         "ports": ["22"]
+    #       }
+    #     ],
+    #     "direction": "INGRESS",
+    #   },
+    #   {
+    #     "name": "firewall2",
+    #     "network": "https://www.googleapis.com/compute/v1/projects/yourproject/global/networks/default",
+    #     "priority": 1000,
+    #     "sourceRanges": [
+    #       "10.0.0.0/8",
+    #       ],
+    #     "allowed": [
+    #       {
+    #         "IPProtocol": "tcp",
+    #         "ports": ["80", "443"]
+    #       }
+    #     ],
+    #     "direction": "INGRESS",
+    #   },
+    # ]
+    # Example firewall rule that WON'T trigger:
+    # [
+    #   {
+    #     "name": "firewall1",
+    #     "network": "https://www.googleapis.com/compute/v1/projects/yourproject/global/networks/default",
+    #     "priority": 1000,
+    #     "sourceRanges": [
+    #       "10.0.0.0/8",
+    #       ],
+    #     "allowed": [
+    #       {
+    #         "IPProtocol": "tcp",
+    #         "ports": ["22"]
+    #       }
+    #     ],
+    #     "direction": "INGRESS",
+    #   },
+    #   {
+    #     "name": "firewall2",
+    #     "network": "https://www.googleapis.com/compute/v1/projects/yourproject/global/networks/default",
+    #     "priority": 1000,
+    #     "sourceRanges": [
+    #       "10.0.0.0/8",
+    #       ],
+    #     "targetTags": ["http-server", "https-server"],
+    #     "allowed": [
+    #       {
+    #         "IPProtocol": "tcp",
+    #         "ports": ["80", "443"]
+    #       }
+    #     ],
+    #     "direction": "INGRESS",
+    #   },
+    # ]
+  - rule_id: 'golden_policy'
+    description: Instances with internal tag must have a specific policy
+    mode: matches
```

Comment:
How is the ```matches``` mode different from ```whitelist``` mode?  Seems to work the same?

Reply:
Whitelist allows flexibility in what rules are implemented. 

A whitelist that says you can have a policy that allows 80, 443, 22, and 3389 to x targets from 0.0.0.0/0 would allow any policy that is a subset of that. That is, the policy could expose 80 to 5.5.5.5 or only 22 to 0.0.0.0/0.

A matches policy requires that the policy is exactly what you would expect. This is useful because you can have a golden policy (like one the enforcer would set) and expect exactly that. If it differs, that is a RuleViolation.